### PR TITLE
return S0 from dki fit

### PR DESCRIPTION
**Pull request by samcoveney**
DKI fitting was not allowing to save S0 so I have made some updates.

### CODE REVIEW

**Comment by skoudoro** [comment]
Thank you for this. Yes, errors are not related, not sure what happens with the CI's and last pytest release.

Looking into it

**Comment by samcoveney** [comment]
Seeing a lot of testing bugs locally:

```
coverage run --source=dipy.reconst.dki -m pytest -s --doctest-modules --verbose ~/Software/dipy_fork/dipy/reconst/tests/test_dki.py
```

It is as if the ```setup_module()``` function is not run, because many tests fail with error:

```
>       bvals = gtab.bvals.copy()
E       AttributeError: 'NoneType' object has no attribute 'bvals'
```

Any idea? Are the tests just wrong now?

**Comment by skoudoro** [comment]
Yes, I do not know what is going on. I agree, it is like `setup_module` stop to work.

I do not succeed to reproduce it locally, I tried on macOS, linux windows. 

 I am very happy that you can reproduce it locally. Can you share you environment (python version, pytest version, etc....)? it will help me to fix this.

Thanks in advance @samcoveney

**Comment by samcoveney** [comment]
Ubuntu 20.04.6 LTS
Python 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0] on linux
pytest.__version__ '8.0.2'
coverage.__version__ '7.4.3'

**Comment by skoudoro** [comment]
Thank you very much! 

I can reproduce it now. It seems that the option `--doctest-modules` make everything fails. there is no problem without this option. I will look deeper

**Comment by skoudoro** [comment]
Also this is specific with pytest 8.0.2, it works with pytest 8.0.0

**Comment by samcoveney** [comment]
In that case, I will consider downgrading pytest to finish off these tests for this pull request, but I'll assume everything will be on hold until you have committed a fix?

**Comment by samcoveney** [comment]
Great!

The issue with ```setup_module``` persists for me though, do you recommend for me to downgrade pytest for now?

**Comment by skoudoro** [comment]
> The issue with setup_module persists for me though, do you recommend for me to downgrade pytest for now?

yes, I will do the same in this repo until there is a fix

**Comment by samcoveney** [comment]
downgraded pytest to 8.0.0 (and not using `--doctest-modules`) and the tests in test_dki.py pass now

**Comment by skoudoro** [comment]
It should be fixed on master, Can you rebase or merge this branch ?

**Comment by samcoveney** [comment]
will rebase now - we'll see if I had done something else wrong...

**Comment by skoudoro** [comment]
it seems many tests are failing @samcoveney. Can you have a look:
```
=========================== short test summary info ============================
FAILED reconst/tests/test_cti.py::test_cti_prediction - ValueError: shapes (81,43) and (42,) not aligned: 43 (dim 1) != 42 (dim 0)
FAILED reconst/tests/test_cti.py::test_cti_fits - ValueError: operands could not be broadcast together with shapes (20,) (21,)
FAILED reconst/tests/test_dki_micro.py::test_single_fiber_model - IndexError: index 5 is out of bounds for axis 0 with size 5
FAILED reconst/tests/test_dki_micro.py::test_wmti_model_multi_voxel - IndexError: index 5 is out of bounds for axis 3 with size 5
FAILED reconst/tests/test_dki_micro.py::test_dki_micro_predict_single_voxel - ValueError: shapes (130,6) and (5,) not aligned: 6 (dim 1) != 5 (dim 0)
FAILED reconst/tests/test_dki_micro.py::test_dki_micro_predict_multi_voxel - ValueError: shapes (130,6) and (5,) not aligned: 6 (dim 1) != 5 (dim 0)
FAILED reconst/tests/test_dki_micro.py::test_dki_micro_awf_only - IndexError: index 27 is out of bounds for axis 3 with size 27
====== 7 failed, 998 passed, 60 skipped, 3 warnings in 1837.12s (0:30:37) ======
```

**Comment by skoudoro** [comment]
Hi @samcoveney,

I plan to release this week, Do you think you will have time to look at this ?

**Comment by samcoveney** [comment]
Completely rewrote it, in style of dti.py, and fixed the pep8 issues.

For the record, it was a bad idea to write dki.py without the ability to save the S0 information!

File: dipy/reconst/dki.py [hunk 1]
```diff
@@ -1597,6 +1598,8 @@ def __init__(self, gtab, fit_method="WLS", *args, **kwargs):
             callable has to have the signature:
                 fit_method(design_matrix, data, *args, **kwargs).
             Default: "WLS"
+        return_S0_hat : bool
```
Comment (skoudoro) [inline dipy/reconst/dki.py:1601]: `bool, optional` instead

File: dipy/reconst/dki.py [hunk 2]
```diff
@@ -1676,8 +1680,14 @@ def fit(self, data, mask=None):
...
-                                 *self.args, **self.kwargs)
-        return DiffusionKurtosisFit(self, params)
+                                        return_S0_hat=self.return_S0_hat,
+                                        *self.args, **self.kwargs)
+
+        S0_params = None
+        if self.return_S0_hat:
+            params, S0_params = params
+
+        return DiffusionKurtosisFit(self, params, S0_params)
```
Comment (skoudoro) [inline dipy/reconst/dki.py:1690]: Can you do `model_S0=S0_params`?  

In april, we will force the codebase to be keyword-only friendly. So this is just to anticipate that.

File: dipy/reconst/dki.py [hunk 3]
```diff
@@ -1731,15 +1746,19 @@ def __init__(self, model, model_params):
...
-        All parameters estimated from the diffusion kurtosis model.
-        Parameters are ordered as follows:
+            All parameters estimated from the diffusion kurtosis model,
+            not including S0.
+            Parameters are ordered as follows:
             1) Three diffusion tensor's eigenvalues
             2) Three lines of the eigenvector matrix each containing the
                 first, second and third coordinates of the eigenvector
             3) Fifteen elements of the kurtosis tensor
+        model_S0 : ndarray (x, y, z, 1) or (n, 1)
```
Comment (skoudoro) [inline dipy/reconst/dki.py:1756]: `, optional`

File: dipy/reconst/dki.py [hunk 4]
```diff
@@ -1731,15 +1746,19 @@ def __init__(self, model, model_params):
...
+            S0 estimated from the diffusion kurtosis model.
 
         """
-        TensorFit.__init__(self, model, model_params)
+
+        TensorFit.__init__(self, model, model_params, model_S0)
```
Comment (skoudoro) [inline dipy/reconst/dki.py:1761]: same comment as above

**Review by skoudoro — COMMENTED** [review]
Hi @samcoveney,

See below my quick comments. It would be great to add some tests also. 

I need to try locally. After the tests addition, it should be ready to go

File: dipy/reconst/dki.py (hunk 2 above)
Comment (samcoveney) [reply dipy/reconst/dki.py:1690]: is this because `model_S0` is a keyword param, and we want to make that clear here?

Comment (skoudoro) [reply dipy/reconst/dki.py:1690]: yes, and we will do it for the whole codebase and examples starting in April.

We plan to follow https://peps.python.org/pep-3102/

**Review by skoudoro — APPROVED** [review]
Ok, It looks good. Thank you for this update @samcoveney 

Merging.

ps: I sent you an email on your Leeds edu address. Let me know if you did not received it

**Pull request merged**